content/blog/2022/2022-03-31-be-informed-with-lombok.md

```diff
@@ -0,0 +1,474 @@
+2. Further, if we **call toString() on the entity outside the Transaction scope**, it could lead to `LazyInitializationException`.
+3. In the case of associations like @ManyToMany between 2 entities, **logging the entity data could result in evaluating circular references and causing StackOverflowError**. In the example above, the `Book` entity 
+will try to fetch all authors of the book. The `Author` entity in turn will try to find all books of the author. This process will keep repeating until it results in an error.
+
+**2. `Avoid @EqualsAndHashCode :`**
+Lombok uses all non-final attributes to evaluate and override default equals and hashCode. This isn't always desirable in the case of entities due to the following reasons:
+1. Most primary keys in the database are **auto-generated** (either using sequences or UUID) 
+during insertion. This **can cause issues in the hashCode computation process**
+as the `ID` is not available beforehand causing unexpected results.
+2. **Every database record is uniquely identified by its primary key**. In such cases using the Lombok implementation of `@EqualsAndHashCode` might not be ideal. 
+
+Although Lombok allows us to include and exclude attributes, for the sake of brevity it might be a **better option to 
+override these methods (toString(), equals(), hashcode()) ourselves** and not rely on Lombok.
+
+### Lombok Annotations hide violations
+Consider a model class in the example below:
+ ````java
+    @Data
+    @Builder
+    @AllArgsConstructor
+    public class CustomerDetails {
+    
+        private String id;
+        private String name;
+        private String buildingNm;
+        private String blockNo;
+        private String streetNm;
+        private String city;
+        private int postcode;
+        private String state;
+        private String country;
+        private Gender gender;
+        private String dateOfBirth;
+        private String email;
+        private String phoneNo;
+        private String drivingLicenseNo;
+        private String licenseIssueState;
+    }
+ ````
+The above class has the following **flaws**:
+1. It is a **poorly designed model class**. We could consider restructuring the fields to create separate classes to store Address and LicenseDetails.
+2. **@Builder hides the complexity of autowiring instances** especially if custom objects 
+having multiple dependencies are added to the class. This affects code readability and destroys the concept of the Single Responsibility Principle.
+3. If you have an application that uses **checkstyle/Sonar**, using these **annotations can hide errors and warnings** which is not desirable.
+This is shown in the sample below:
+   {{% image alt="delombok example" src="images/posts/lombok/checkstyle.png" %}}
+   {{% image alt="delombok example" src="images/posts/lombok/checkstyle_err.png" %}}
+4. With **`@AllArgsConstructor`**  referring to multiple same-type parameters, it is easy to accidentally define parameters out of order. 
```

Comment:
This doesn't have to do with "hiding validations" (which is the title of this section), does it? It feels like this point should move into the next section, because it's about a dependency to generated code.